presto-parquet/src/main/java/io/prestosql/parquet/predicate/TupleDomainParquetPredicate.java

```diff
@@ -389,4 +502,353 @@ private static ParquetCorruptionException corruptionException(String column, Par
+            if (columnDomain.intersect(domain).isNone()) {
+                return true;
+            }
+            return false;
+        }
+    }
+
+    class ColumnIndexValueConverter
+    {
+        private final Map<String, Function<Object, Object>> ciConversions;
+
+        private ColumnIndexValueConverter(List<RichColumnDescriptor> columns)
+        {
+            this.ciConversions = new HashMap<>();
+            for (RichColumnDescriptor column : columns) {
+                ciConversions.put(column.getPrimitiveType().getName(), getColumnIndexConversions(column.getPrimitiveType()));
+            }
+        }
+
+        public List<Long> getMinValuesAsLong(Type type, ColumnIndex columnIndex, String column)
+        {
+            int pageCount = columnIndex.getMinValues().size();
+            List<ByteBuffer> minValues = columnIndex.getMinValues();
+            List<Long> mins = new ArrayList<>();
+            for (int i = 0; i < pageCount; i++) {
+                if (TINYINT.equals(type) || SMALLINT.equals(type) || INTEGER.equals(type)) {
+                    int minValue = converter.convert(minValues.get(i), column);
+                    mins.add((long) minValue);
+                }
+                else if (BIGINT.equals(type)) {
+                    long minValue = converter.convert(minValues.get(i), column);
+                    mins.add(minValue);
+                }
+                else if (REAL.equals(type)) {
+                    int minValue = floatToRawIntBits(converter.convert(minValues.get(i), column));
+                    mins.add((long) minValue);
+                }
+            }
+            return mins;
+        }
+
+        public List<Long> getMaxValuesAsLong(Type type, ColumnIndex columnIndex, String column)
+        {
+            int pageCount = columnIndex.getMaxValues().size();
+            List<ByteBuffer> maxValues = columnIndex.getMaxValues();
+            List<Long> maxs = new ArrayList<>();
+            if (TINYINT.equals(type) || SMALLINT.equals(type) || INTEGER.equals(type)) {
+                for (int i = 0; i < pageCount; i++) {
+                    int maxValue = converter.convert(maxValues.get(i), column);
+                    maxs.add((long) maxValue);
+                }
+            }
+            else if (BIGINT.equals(type)) {
+                for (int i = 0; i < pageCount; i++) {
+                    long maxValue = converter.convert(maxValues.get(i), column);
+                    maxs.add((long) maxValue);
+                }
+            }
+            else if (REAL.equals(type)) {
+                for (int i = 0; i < pageCount; i++) {
+                    int maxValue = floatToRawIntBits(converter.convert(maxValues.get(i), column));
+                    maxs.add((long) maxValue);
+                }
+            }
+            //TODO: Else
+            return maxs;
+        }
+
+        public List<Double> getMinValuesAsDouble(Type type, ColumnIndex columnIndex, String column)
```

Comment:
Most of these `get{Min,Max}ValuesAs[Type]` methods should be folded into a generic form to go with the suggested changes to `getDomain`.

I think you'll want something that looks mostly like this (but you may need to replace `getJavaType` to make it work).

```java
        public List<?> getMinValues(Type type, ColumnIndex columnIndex, String column)
        {
            return columnIndex.getMinValues().stream()
                    .map(value -> converter.convert(value, column))
                    .map(type.getJavaType()::cast)
                    .collect(Collectors.toList());
        }
```